elasticmagic-query-filters/src/commonMain/kotlin/dev/evo/elasticmagic/qf/FacetExpressionsFilter.kt

```diff
@@ -0,0 +1,112 @@
+
+class FacetExpressionFilterResult(
+    override val name: String,
+    override val paramName: String,
+    val results: List<FaceExpressionValue>,
+    val mode: FilterMode
+) : FilterResult
+
+class PreparedFacetExpressionFilter(
+    name: String,
+    paramName: String,
+    private val selectedFilterExprs: List<ExpressionValue>,
+    private val facetFilterExprs: List<ExpressionValue>,
+    val mode: FilterMode,
+) : PreparedFilter<FacetExpressionFilterResult>(
+    name,
+    paramName,
+    getFacetFilterExpr(selectedFilterExprs.map { it.expr }, mode)
+) {
+    private val filterAggName = "qf.$name.filter"
+
+    private fun makeAggName(value: String) = "qf.$name:$value"
+
+    override fun apply(
+        searchQuery: SearchQuery<*>,
+        otherFacetFilterExpressions: List<QueryExpression>
+    ) {
+
+        val filterAggs = facetFilterExprs.associate { makeAggName(it.name) to FilterAgg(it.expr) }
+        val filters =
+            if (mode == FilterMode.INTERSECT) (otherFacetFilterExpressions + facetFilterExpr).mapNotNull { it }
+            else otherFacetFilterExpressions
+
+        val aggs = if (filters.isNotEmpty()) {
+            val filtered = filters.filter { f -> facetFilterExprs.all { it.name != f.name } }
+            mapOf(
+                filterAggName to FilterAgg(
+                    maybeWrapBool(Bool::must, filtered),
+                    aggs = filterAggs
+                )
+            )
+        } else filterAggs
+
+        searchQuery.aggs(aggs)
+        if (facetFilterExpr != null) {
+            searchQuery.postFilter(facetFilterExpr)
+        }
+    }
+
+    override fun processResult(
+        searchQueryResult: SearchQueryResult<*>,
+    ): FacetExpressionFilterResult {
+        val selectedNames = selectedFilterExprs.map { it.name }
+        val agg = searchQueryResult.aggs[filterAggName]
+
+        val parsedAggs = if (agg != null)
+            (agg as SingleBucketAggResult).aggs
+        else
+            searchQueryResult.aggs
+
+        val results = facetFilterExprs.map { facetFilterExpr ->
+            val filterAgg = parsedAggs[makeAggName(facetFilterExpr.name)] as LongValueAggResult
+            FaceExpressionValue(
+                facetFilterExpr.name,
+                selectedNames.contains(facetFilterExpr.name),
+                filterAgg.value
+            )
+        }
+
+        return FacetExpressionFilterResult(name, paramName, results, mode)
+    }
+}
+
+open class FacetExpressionsFilter(
+    name: String? = null,
+    private val allValues: List<ExpressionValue> = emptyList(),
+    val mode: FilterMode = FilterMode.UNION,
+) : Filter<FacetExpressionFilterResult>(name) {
+
+    override fun prepare(name: String, paramName: String, params: QueryFilterParams): PreparedFacetExpressionFilter {
+        val filterValues = params.getOrElse(listOf(paramName)) { null } ?: return PreparedFacetExpressionFilter(
```

Comment:
I would prefer to have a single construction site of `PreparedFacetExpressionFilter` instance. We only need to find out `selectedValues`